File: android/beagle/src/main/java/br/com/zup/beagle/android/imagedownloader/DefaultImageDownloader.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.imagedownloader

import android.graphics.Bitmap
import android.widget.ImageView
import androidx.lifecycle.lifecycleScope
import br.com.zup.beagle.android.cache.imagecomponent.ImageDownloader
import br.com.zup.beagle.android.data.formatUrl
import br.com.zup.beagle.android.logger.BeagleMessageLogs
import br.com.zup.beagle.android.utils.CoroutineDispatchers
import br.com.zup.beagle.android.widget.RootView
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

internal class DefaultImageDownloader : BeagleImageDownloader {

    private val imageDownloader: ImageDownloader = ImageDownloader()

    override fun download(url: String, imageView: ImageView, rootView: RootView) {
        imageView.post {
            rootView.getLifecycleOwner().lifecycleScope.launch(CoroutineDispatchers.IO) {
                try {
                    val formattedUrl = url.formatUrl().takeIf { it.isNotEmpty() } ?: url
                    val bitmap = imageDownloader.getRemoteImage(
                        formattedUrl,
                        imageView.width,
                        imageView.height,
                        rootView.getContext(),
                    )
                    setImage(imageView, bitmap)
                } catch (e: Exception) {
                    BeagleMessageLogs.errorWhileTryingToDownloadImage(url, e)
                }
            }
        }
    }

    private suspend fun setImage(view: ImageView, bitmap: Bitmap?) {
        withContext(CoroutineDispatchers.Main) {
            view.setImageBitmap(bitmap)
        }
    }
}

File: flutter/beagle/lib/default/default_image_downloader.dart
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'dart:io';
import 'dart:typed_data';

import 'package:beagle/interface/beagle_image_downloader.dart';
import 'package:beagle/interface/http_client.dart';
import 'package:beagle/networking/beagle_request.dart';
import 'package:flutter/foundation.dart';

class DefaultBeagleImageDownloader implements BeagleImageDownloader {
  DefaultBeagleImageDownloader({
    @required this.httpClient,
  }) : assert(httpClient != null);

  final HttpClient httpClient;

  @override
  Future<Uint8List> downloadImage(String url) async {
    final request = BeagleRequest(url);
    final response = await httpClient.sendRequest(request);

    if (response.status != HttpStatus.ok) {
      throw BeagleImageDownloaderException(
          statusCode: response.status, url: request.url);
    }

    final bytes = response.bodyBytes;
    if (bytes.lengthInBytes == 0) {
      throw Exception('Image is an empty file: $url');
    }

    return bytes;
  }
}

class BeagleImageDownloaderException implements Exception {
  BeagleImageDownloaderException(
      {@required this.statusCode, @required this.url})
      : assert(url != null),
        assert(statusCode != null),
        _message = 'HTTP request failed, statusCode: $statusCode, $url';

  final int statusCode;

  final String _message;

  final String url;

  @override
  String toString() => _message;
}
